### ISSUE TITLE
defines too long ?

### ISSUE THREAD

**Issue by 3615JMD** [jasonful/lcc#5]
First I want to thank Jason for sharing his work !

I'm working on Windows10 and I eventually could get the ulpcc compiler working, but I had to reduce the length of many #defines in the file rtc_cntl_reg.h

exemple:
#define RTC_CNTL_BROWN_OUT_CLOSE_FLASH_ENA_M  (BIT(14))
#define RTC_CNTL_BROWN_OUT_CLOSE_FLASH_ENA_V  0x1
#define RTC_CNTL_BROWN_OUT_CLOSE_FLASH_ENA_S  14

changed to the following :

#define RTCCNTLBROWNOUTCLOSEFLASH_ENA_M  (BIT(14))		// to make labels shorters
#define RTCCNTLBROWNOUTCLOSEFLASH_ENA_V  0x1
#define RTCCNTLBROWNOUTCLOSEFLASH_ENA_S  14

because otherwise,the compiler appeared to ignore (trim out) the last character(s) and issued a multiple definition error as the beginning of the label was the same for the 3 above defines.

has those defines changed ? have you found a way to have the compiler/preprocessor accept longer labels ? did I missed something ?

best regards,
Jean-Marc.

**Comment by jasonful** [comment]
Hmm, I can't reproduce this.  Can you send me your .c file? It looks like a 32-char limit, but I don't think the lcc preprocessor has such a limit.

**Comment by 3615JMD** [comment]
Hi Jason,

 

Thank you for your reply.

 

The issue appeared when I included the header file containing all the RTC registers:   rtc_cntl_reg.h

 

To make the issue appear Just add 

 

#include <rtc_cntl_reg.h>

 

to the top of any ulp exemple c file

 

In my installation, that header file was here:

 

C:\Users\gqs\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.1\tools\sdk\include\soc\soc\rtc_cntl_reg.h

 

There are a lot of defines such as below:

 

#define RTC_CNTL_PROCPU_STAT_VECTOR_SEL_M  (BIT(13))

#define RTC_CNTL_PROCPU_STAT_VECTOR_SEL_V  0x1

#define RTC_CNTL_PROCPU_STAT_VECTOR_SEL_S  13

 

And the preprocessor seemed not to compare till the end of label.

 

Best regards,

Jean Marc

 

 

 

De : jasonful <notifications@github.com> 
Envoyé : mercredi 12 février 2020 07:47
À : jasonful/lcc <lcc@noreply.github.com>
Cc : Jean-Marc <3615jmd@free.fr>; Author <author@noreply.github.com>
Objet : Re: [jasonful/lcc] defines too long ? (#5)

 

Hmm, I can't reproduce this. Can you send me your .c file? It looks like a 32-char limit, but I don't think the lcc preprocessor has such a limit.

**Comment by jasonful** [comment]
I still can't reproduce this.  Can you verify that the version of cpp.exe you are running comes from %ULPCCDIR%\bin ?  Or maybe produce a minimal .c file that illustrates the error?

**Comment by 3615JMD** [comment]
Hi Jason,

 

Any C file will create the issue here.

 

You don’t even need to include rtc_cntl_reg.h because it is included from the compulsory ulp_c.h

 

I made a short video showing the problem. You can see it here:     https://we.tl/t-w9UnBmygwF

 

Regarding cpp.exe it seems that lcc is not using the one that is in the same directory ?!!!   

I have renamed for a while that  cpp.exe to cpp2.exe and lcc works just the same.

 

I have checked the path and potentially there are many other cpp.exe on my system (Borland, Microchip, Arduino…)

 

Then I have cleared the path totally and I get this error:     lcc: cpp: No such file or directory

 

It shows that lcc is not even looking in the same directory to find cpp.exe !!!

 

Then  I have set the path to only  “C:\PROJETS\ESP32\ULPCC\bin”

 

And now lcc compiles the test program even with the long defines labels…    YES !

 

The issue is then solved for manual compiling, but as I have integrated the ulpcc compiler into the Arduino (after modifying some files because windows was not supported) I shall now investigate how to modify the path automatically in the compile process…

 

Thanks for your help !    (but I really don’t understand why lcc does not call the cpp that is in the same directory that would be so simple !)

 

May be I need to create an new environment variable pointing to the correct cpp.exe ?

Best regards,

Jean Marc.

 

 

 

De : jasonful <notifications@github.com> 
Envoyé : mercredi 12 février 2020 23:52
À : jasonful/lcc <lcc@noreply.github.com>
Cc : Jean-Marc <3615jmd@free.fr>; Author <author@noreply.github.com>
Objet : Re: [jasonful/lcc] defines too long ? (#5)

 

I still can't reproduce this. Can you verify that the version of cpp.exe you are running comes from %ULPCCDIR%\bin ? Or maybe produce a minimal .c file that illustrates the error?

**Comment by jasonful** [comment]
Yes, lcc pulls cpp.exe from your path.  The code was designed for a Unix system where lcc would be installed in some fixed location.  I actually had to make a change to get it to use the path, but perhaps the right answer is to modify etc\win32.c to pull cpp.exe from the directory specified by the environment variable ULPCCDIR (as determined at runtime rather than compile time).

**Comment by jasonful** [comment]
Or I could modify the README.md to ask people to put ULPCCDIR first in the path...

**Comment by 3615JMD** [comment]
And what about trying first to find it in the same directory from where the lcc.exe was launched ?

Because putting ULPCCDIR first in the path might wreck other compilers ?

 

Good news:  Arduino calls the right cpp.exe  so everything is fine now for me.

(I’ll check later how does the script find the right cpp)

 

In summary I had the issue because I tried to get ulpcc working independently before trying to have it integrated into the Arduino env. And because several previous cpp installations existed.

 

Thanks for your help.

Best regards,

JM

 

 

PS:   FYI since then I have found a bug in the ulp assembler (wrong limitation of jumps above address 2048) which I have reported.  Only appears on large programs, which was the case for me.

 

Here:      https://github.com/espressif/binutils-esp32ulp/issues/16

 

 

 

 

De : jasonful <notifications@github.com> 
Envoyé : jeudi 13 février 2020 17:48
À : jasonful/lcc <lcc@noreply.github.com>
Cc : Jean-Marc <3615jmd@free.fr>; Author <author@noreply.github.com>
Objet : Re: [jasonful/lcc] defines too long ? (#5)

 

Yes, lcc pulls cpp.exe from your path. The code was designed for a Unix system where lcc would be installed in some fixed location. I actually had to make a change to get it to use the path, but perhaps the right answer is to modify etc\win32.c to pull cpp.exe from the directory specified by the environment variable ULPCCDIR (as determined at runtime rather than compile time).
